en/lessons/basics/pattern-matching.md

```diff
@@ -1,107 +1,283 @@
 
 {% include toc.html %}
 
-## Match Operator
+## Extensions and Imports
+
+You will need to add the following lines to the top of your Haskell source 
+file (`.hs`) to be able to run these examples
+
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+import Data.Text ( Text )
+```
+
+## Enumerated Type Example
+
+Say we have a type for the four suits of playing cards
+
+```haskell
+data Suit = Hearts | Diamonds | Clubs | Spades
+```
+
+### Using a Case Statement
+
+Let's write a case statement which will read the color of the suit
+
+```haskell
+-- Show the color of the suit
+suitColorCase :: Suit -> Text
+suitColorCase suitIn = case suitIn of
+     Hearts   -> "Red"
+     Diamonds -> "Red"
+     Clubs    -> "Black"
+     Spades   -> "Black"
+```
+
+The case statement is using pattern matching to decide which text to return
+
+The value of `suitIn` is evaluated, and then checked against the given patterns
+in the case statement
+- In this example, each pattern is just a data constructor
+- We will get to see more sophisticated patterns a bit later
+
+### Using Pattern Matching
+
+It turns out that we can write the same function more simply
+
+```haskell
+-- Show the color of the suit
+suitColor :: Suit -> Text
+suitColor Hearts   = "Red"
+suitColor Diamonds = "Red"
+suitColor Clubs    = "Black"
+suitColor Spades   = "Black"
+```
+
+Here we have built the patterns into the function definition, without using a
+case statement
+
+The function works the same way: any input value is inspected and matched
+with the patterns provided for the function definition
+
+In truth, the case statement is redundant in the first example, and as
+Haskell programmers we like to keep our code as concise as is reasonable
+
+This can feel awkward at first, since in imperative languages there is usually
+only a single declaration of a function
+- In Haskell, pattern matching syntax is common
+- Always keep in mind that Haskell functions are definitions, and with pattern
+  matching we are providing different definitions for different inputs
+
+## Pattern Matching Syntax
+
+```
+functionName (pattern1) = (expression1)
+functionName (pattern2) = (expression2)
+...
+functionName _ = (expressionLast)
+```
+
+- A function is defined more than one time, each with a different pattern
+- For the first pattern that matches the value with which the function was
+  called, its expression is evaluated
+- The underscore (`_`) is a catch-all that matches any value
+
+### Patterns In General
+
+A pattern is a combination of data constructors that could match a value
+
+Functions are not permitted in patterns -- If you need to use a function,
+consider using guards
+
+#### Examples of Patterns
+
+```
+"banana" -- A `Text` value
+7        -- An `Int` value
+[]       -- An empty list
+(x:xs)   -- A list with one or more values, constructed with `:`
+         -- `x` is the first element of the list, and `xs` is the rest of the list
+Nothing  -- A constructor for the `Maybe` type
```

Comment:
`Maybe` is not a type. `Maybe a` is.

Reply:
`Maybe` is very much a type, of kind `* -> *`. `Maybe a` is a type of kind `*`. Being a higher-kinded type doesn't make it less of a type.

Reply:
What I mean to say is that since `Nothing :: Maybe a`, the correct thing to say is that `Nothing` is a constructor for the `Maybe a` datatype.

Reply:
Decided to leave this as-is